src/main/java/org/openpnp/machine/reference/ReferencePnpJobProcessor.java

```diff
@@ -816,12 +816,35 @@ protected void doReset() throws Exception {
      */
     protected void doSkip() throws Exception {
         if (plannedPlacements.size() > 0) {
-            PlannedPlacement plannedPlacement = plannedPlacements.remove(0);
-            JobPlacement jobPlacement = plannedPlacement.jobPlacement;
-            Nozzle nozzle = plannedPlacement.nozzle;
-            discard(nozzle);
-            jobPlacement.status = Status.Skipped;
-            Logger.debug("Skipped {}", jobPlacement.placement);
+
+            // iterate through planned placement in this cycle (number of planned placements ==
+            // number of nozzles)
+            for (PlannedPlacement plannedPlacement : plannedPlacements) {
+                JobPlacement jobPlacement = plannedPlacement.jobPlacement;
+
+                if (plannedPlacement.stepComplete) {
+                    // go over placements having the current step already completed
+                    continue;
+                }
+
+                // remove the placement to be skipped from list
+                plannedPlacements.remove(plannedPlacements.indexOf(plannedPlacement));
```

Comment:
This is likely to cause a ConcurrentModificationException. You can't modify a collection that you are iterating unless you use an Iterator and the Iterator.remove() method. See https://stackoverflow.com/questions/18448671/how-to-avoid-concurrentmodificationexception-while-removing-elements-from-arr. Other that that, this change looks solid!